File: array/medium/golang_solutions/recursion_backtracking/combinations_77.go
package recursion_backtracking

func combine(n int, k int) [][]int {
	result := make([][]int, 0)
	backtrack(&result, make([]int, 0), k, n, 1)
	return result
}

func backtrack(result *[][]int, temp []int, k, n, start int) {
	if len(temp) == k {
		copyOf := make([]int, k)
		copy(copyOf, temp)
		*result = append(*result, copyOf)
		return
	}
	for i := start; i <= n; i++ {
		temp = append(temp, i)
		backtrack(result, temp, k, n, i+1)
		temp = temp[:len(temp)-1]
	}
}

File: backtracking_recursion/java_solutions/Combinations_77.java
package backtracking_recursion.java_solutions;

import java.util.ArrayList;
import java.util.List;

public class Combinations_77 {

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();

        backtrack(result, new ArrayList<>(), n, k, 1);

        return result;
    }

    public void backtrack(List<List<Integer>> result, List<Integer> temp, int n, int k, int start) {
        if (temp.size() == k) {
            result.add(new ArrayList<>(temp));
            return;
        }
        for (int i = start; i <= n; i++) {
            temp.add(i);
            backtrack(result, temp, n, k, i + 1);
            temp.remove(temp.size() - 1);
        }
    }
}
